### PR TITLE
merge: consider type when reducing a volume

### PR DESCRIPTION
**Pull request by aesteve-rh**
When finalizing a merge, only the format (cow) and the `optimal_size` are considered to discern when to reduce a volume. However, calculating the optimal size can be avoided in the case of a preallocated block volume, as it cannot be reduced.

Also, in the StorageDomain, `reduceVolume` can be called with no check at all, which is dangerous.

Use the volume `can_reduce` check before obtaining the optimal size, and before reducing the volume (both in the cold merge and the storage domain), so that we consider the type of volume before reducing it.

Fixes: https://github.com/oVirt/vdsm/issues/185
Bug-Url: https://bugzilla.redhat.com/2081493
Signed-off-by: Albert Esteve <aesteve@redhat.com>

### CODE REVIEW

File: lib/vdsm/storage/merge.py
Comment (nirs) [inline lib/vdsm/storage/merge.py:266]: We need to check if a volume can be reduced also in line 265. If a volume cannot be reduced, there is no need to get the optimal size.

File: tests/storage/merge_test.py
Comment (nirs) [inline tests/storage/merge_test.py:311]: Don't add logic here, the test should configure this properly to test all cases.

File: tests/storage/merge_test.py
Comment (nirs) [inline tests/storage/merge_test.py:466]: This does not test the interesting cases - the issue is with preallocated qcow2 on block, and this tests preallocated raw.

**Review by nirs — CHANGES_REQUESTED** [review]
This conflicts with #133. Lets wait until it is merged. I plan to work on this next week.

File: lib/vdsm/storage/merge.py
Comment (aesteve-rh) [reply lib/vdsm/storage/merge.py:266]: OK.

Why is the optimal size obtained within the context and the shrink called outside the context?

File: lib/vdsm/storage/merge.py
Comment (aesteve-rh) [reply lib/vdsm/storage/merge.py:270]: First point is tackled here: https://github.com/oVirt/vdsm/pull/178

For the rest, I will wait until #133 :)

File: lib/vdsm/storage/merge.py
Comment (nirs) [reply lib/vdsm/storage/merge.py:266]: To compute optimal size we measure the image using "qemu-img measure" - so we must prepare the image (activate the logical volume in block storage). Reducing the image is possible when the logical volume is active, but LVM issues warnings about this, so we prefer to teardown the volume before reducing.

The flow is:
- activate the lv and its parent lv (recursively)
- measure the qocw2 image in the lv
- deactivate the lv and its parent (recursively)
- if the lv is bigger than the optimal size, reduce it.

I guess the code needs a comment explaining this.

Comment (aesteve-rh) [reply lib/vdsm/storage/merge.py:266]: Great, thanks for the explanation :)
I assumed something like that from the code, but a clear explanation is much better! I may add a comment to explain that in the code.

File: tests/storage/merge_test.py
Comment (nirs) [inline tests/storage/merge_test.py:553]: Maybe test_reduce_skipped?

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:1666]: Keeping . at the end of comments will make this even nicer.

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:1695]: I would change this to something like:

    # Volume is not prepared but it does fail since we skip the reduce
    # for preallocated volume.

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:1699]: The TODO: is not needed - we don't want to check that reduce was not called, but that reducing a preallocated volume does not change the volume size, and this check is the
best way to do this.

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:1662]: "over extended volume"

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:1671]: This does not test that we reduced the volume to the optimal size.

To fix this test:
1. Create the volume with initial_size=5 * GiB
2. Check that the volume size after the reduce is vol.optimal_size()

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:1700]: using apparentsize is better. truesize is kind of not working for block storage, the value is always same as apparentsize. In file storage this is the size allocated
(like du /path).

**Review by nirs — COMMENTED** [review]
Nice!

**Comment by nirs** [comment]
This fixes 2 bugs:
- reduce after cold merge - https://bugzilla.redhat.com/2081493
- reduce when calling Volume.reduce - we don't have a bug

It will be useful to add upstream issue to the second fix, since users may hit this issue
and the issue will help them to understand the problem and find the fix.

You should be able to reproduce this issue using the reduce example in the sdk:
https://github.com/oVirt/python-ovirt-engine-sdk4/blob/main/examples/reduce_disk.py

Create preallocated volume enabled for incremental backup. The disk will be created fully allocated. Run the reduce script - without the fix it will be reduced to 2.62g. With the fix it will not be reduced.

File: tests/storage/blocksd_test.py
Comment (aesteve-rh) [reply tests/storage/blocksd_test.py:1666]: Agh, I try because I see all comments with it and want to keep it consistent, but I'm not used to have periods at the end of sentences in comments, so I miss them easily. I'll fix them.

File: tests/storage/blocksd_test.py
Comment (aesteve-rh) [reply tests/storage/blocksd_test.py]: I was not sure since in merge_tests the call is actually checked using the `__calls__` attribute, using fake volumes, and hence the comment. I will remove it.

File: tests/storage/blocksd_test.py
Comment (nirs) [inline tests/storage/blocksd_test.py:1675]: This is valid python, but harder to read. It will be nicer as:

    new_size = dom.getVolumeSize(img_uuid, vol_id).apparentsize
    assert new_size == vol.optimal_size()

**Review by nirs — DISMISSED** [review]

**Review by nirs — APPROVED** [review]

**Pull request merged**
